### ISSUE TITLE
Crash - Removing CCSprite or Replace Scene [Duplicate #402]

### ISSUE THREAD

**Issue by mikeytdan** [spritebuilder/SpriteBuilder#431]
Original Post: http://forum.spritebuilder.com/t/crash-removing-sprites/1001

Develop Branch

[EDIT] - Potential Fix

I believe the issue causing the crash is the CCActionCallFunc. the CCActionCallFunc target is nil. Here is my temp fix in CCActionInterval.m:

```
// New action. Start it.
if( found != _last ) {
    CCActionCallFunc *callFunction = nil;
    if(found == 1 && [_actions[found] isKindOfClass:[CCActionCallFunc class]]) {
        callFunction = (CCActionCallFunc *)_actions[found];
    }
    if(!callFunction || (callFunction && callFunction.target)) {
        [_actions[found] startWithTarget:_target];
    }
}
```

[/Edit]

Ways to Duplicate: 

Create CCSprites using CCBReader and add them to a scene then remove them. The app will eventually crash.

Create a scene with a button and load a different scene:

```
- (void)pianoKeyPressed {
    [[CCDirector sharedDirector] replaceScene:[CCBReader loadAsScene:@"MusicScene"]];
}
```

More info:

I've been running into this issue and don't know if it's crashing because I'm doing something wrong. When I remove CCSprites from the scene, stop all their actions, and get rid of any pointers to the CCSprite it causes the app to crash in the CCAction.m file on the following method:

```
-(void) startWithTarget:(id)aTarget
{
    _originalTarget = _target = aTarget;
}
```

It crashes on the line where the method name is defined. "-(void) startWithTarget: (id)aTarget"

In the stack tree it's hitting the following lines in CCActionInterval.m:

```
if( found != _last )
    [_actions[found] startWithTarget:_target];
```

[EDIT] - More Info:

The actions in the _actions list are CCActionDelay and CCActionCallFunc

I believe I know the issue but I don't know how to fix it. I believe it's crashing because it's trying to run the animation on the object being removed. The stopAllActions method doesn't stop the animation system set up in SpriteBuilder. I change the default animation time of 10 seconds to 1 and added a delay to remove the sprite after 1.2 seconds. It causes the app not to crash but it's a hacky workaround.

Long Story Short: Removing a CCSprite will cause a crash unless you remove it with a delay that is bigger than it's animation time.

[/Edit]

I don't know why this is happening or I'd fix it myself. I tried changing the if( found != last ) to if( found != last && target && actions[found]) but that had no effect. I also tried checking it the _target was not equal to NSNull but that caused a crash on the line I defined it on.

Any help/input on this issue would be helpful it's causing me a lot of headache.

**Comment by cocojoe** [comment]
Oh I'm guessing this is just #402 It is fixed in a pull on cocos2d/develop branch. Could if you could check though just in case.

**Comment by mikeytdan** [comment]
Looks like that was it, thanks!
